### PR TITLE
feat(common/models): lexicon traversal

### PR DESCRIPTION
**Pull request by jahorton**
While our currently-published model template is a Trie-based wordlist, it's important to note that we may support other model templates in the future.  When it comes to examining the lexicon for possible words, the important thing for an edit-distance based search is that we can search the lexicon '_like_' a Trie, not necessarily that the full model itself is Trie-based.

For highly agglutinative languages, it may be more efficient to reconstruct the words from morphemes (as needed) than to actually store every possible word within a Trie.  Furthermore, edit distance doesn't care whether or not a model's backing Trie stores probabilities on a per-word basis or uses some alternate function.  It only cares what words 'neighbor' the current word it's examining.

----

Past that... it turns out that even our current Trie-wordlist's Trie storage's interface doesn't quite match what I'm wanting for use with my edit-distance search plans.

For example, suppose a user has typed "trou" and is looking to type the word "trouble."  With the super-basic test model (See `english-1000.json`), we have the node for "tro"... that jumps straight to the leaf node for "trouble".  For searching, it's best if we have a "virtual" node for "trou".
- "trou" vs "tro":  edit distance 1
- "trou" vs "trou": edit distance 0
- "trou" vs "trouble": edit distance 3

That same model also supports:
- "trou" vs "tru": edit distance 1
- "trou" vs "true": edit distance 2
- "trou" vs "truth": edit distance 3

If a "trou" node is not available, any suggestions prefixed by either "tro" or "tru" will be given equal weight.  With a "trou" node, we know to prioritize "trou" entries over "tru" entries due to the edit distance difference.  

To be a bit more explicit about my thought process:
- Once a compatible edit-distance based search is available, the original `predict()` method would be used to find suggestions based on the closest-matching prefixes.
    - That is, edit-distance search is largely for 'correction'.
    - Of course, if a prefix is a word itself, that's great!
- For the average case, we're not likely to compute edit distance against all common prefixes / nodes of a Trie, let alone every word in the lexicon.
    - If we find that "edit distance = 1" provides 12+ suggestions and everything else left to search is at least edit distance 2... why search further?
    - I may add an extra `LexiconTraversal` field:  `representedCount`.  This would help the search to track how many suggestions are already available without the need to even call `predict`.
    - That said, it may be the case that `representedCount`'s only uses would be at times we'd look to call `predict` anyway, so... I'm holding off on that for now.
- Between inputs, we would save all precomputed intermediate calculations.  Part of this is _also_ saving the traversal state.
    - These would be fed into a priority queue to perform breadth-first search... and without a need to re-traverse the lexicon unnecessarily.

-------

Finally... please feel free to critique the nomenclature this PR introduces.  I struggled to get it to its current state, and I'm still not entirely happy with the current terminology.

### CODE REVIEW

**Comment by mcdurdin** [comment]
I think I might let @eddieantonio review this one but one small request: please make sure there are some good code comments for future maintainers -- perhaps turning parts of the excellent intro to this PR into comments where appropriate 😁

**Comment by jahorton** [comment]
I made sure to document the core aspects of the `LexiconTraversal` interface already:

https://github.com/keymanapp/keyman/blob/ceba91fab6ea16498a2592cfbacc2a145d59726c/common/models/types/index.d.ts#L20-L38

Much of the other detail in the 'description' is documentation about my current edit-distance search design plans, which likely would belong elsewhere.  I felt the additional details were necessary here for context because the algorithm the PR's contents will facilitate is not yet implemented.

File: common/models/templates/src/trie-model.ts [hunk 1]
```diff
@@ -161,6 +161,116 @@
...
+
+      *children(): Generator<{char: string, traversal: () => LexiconTraversal}> {
+        let root = this.root;
+
+        if(root.type == 'internal') {
+          for(let i = 0; i < root.values.length; i++) {
+            let entry = root.values[i];
+            let entryNode = root.children[entry];
+
+            // SMP check
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:186]: Pedantic note: "SMP" standard for the [Supplementary Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Supplementary_Multilingual_Plane), which is U+010000–U+01FFFF. This check is for finding a surrogate pair, which may encode any code point outside of the Basic Multilingual Plane (U+0000–U+FFFF).

All code points U+010000 and above don't officially have a better name than "non-BMP", but I think they're colloquially called the "[astral planes](http://www.opoudjis.net/unicode/unicode_astral.html)", as UTF-16 must astral project to encode code points in the range 😂

File: common/models/templates/src/trie-model.ts [hunk 2]
```diff
@@ -161,6 +161,116 @@
...
+            let charCode = entry.charCodeAt(0);
+            if(charCode >= 0xD800 && charCode <= 0xDBFF) {
+              // First part of a SMP char.
+              // For now, we'll just assume the second always completes such a char.
+              //
+              // Note:  Things get nasty here if this is only sometimes true; in the future,
+              // we should compile-time enforce that this assumption is always true if possible.
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:193]: TODO: always ensure every string in the models are USV strings (unicode scalar values); i.e., each surrogate pair is composed of exactly a high surrogate followed by a low surrogate.

File: common/models/types/index.d.ts [hunk 3]
```diff
@@ -78,6 +109,13 @@ declare interface LexicalModel {
    * @see LexicalModelPunctuation
    */
   readonly punctuation?: LexicalModelPunctuation;
+
+  /**
+   * Lexical models _may_ provide a LexiconTraversal object usable to enhance 
+   * prediction and correction results.  The returned object represents the
+   * unfiltered lexicon (with an empty prefix).
+   */
+  getRootTraversal?(): LexiconTraversal;
```
Comment (eddieantonio) [inline common/models/types/index.d.ts:118]: Using my suggestion from before...

```suggestion
  traverseFromRoot?(): LexiconTraversal;
```

File: common/models/types/index.d.ts [hunk 4]
```diff
@@ -17,6 +17,37 @@
...
+  /**
+   * Provides an iterable pattern used to search for words with a prefix matching
+   * the current traversal state's prefix when a new character is appended.
+   * 
+   * For example, if the current traversal state corresponds to 'th', children() may return
+   * an iterator with states corresponding 'e' (for 'the', 'then', 'there'),
+   * 'a' (for 'than', 'that'), etc.
+   * 
+   * @param char      The character suffixed to the existing lookup-prefix's string after
+   *                  any 'keying' transformations are applied.
```
Comment (eddieantonio) [inline common/models/types/index.d.ts:34]: This description is very hard for me to follow. Could you maybe rewrite it?

I think it's saying that, if the traversal is active for `th`, then `char` will be `"a"` or `"e"`. I'm not sure how to succinctly describe this...

Another thing: I know `USVString` implies it, but could you please note that `char` can be an astral code points may, with a JavasScript `.length` of 2?

File: common/models/types/index.d.ts [hunk 5]
```diff
@@ -17,6 +17,37 @@
...
+   * @param traversal A closure providing an iterable over the possible child states
+   * of the resulting state.
```
Comment (eddieantonio) [inline common/models/types/index.d.ts:36]: Call this to traverse the node corresponding to `char`?

File: common/models/templates/src/trie-model.ts [hunk 6]
```diff
@@ -161,6 +161,116 @@
...
+          let matches = this.root.entries.filter(function(entry) {
+            return entry.key == prefix;
+          })
+          if(matches.length > 0) {
+            return matches.map(function(value) { return value.content });
+          } else {
+            return undefined;
+          }
+        } else {
+          let matchingLeaf = this.root.children['\uFDD0'];
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:265]: TODO: extract the constant `\uFDD0` into a constant... maybe with a better name than this: https://github.com/keymanapp/keyman/blob/df283e6fd7e857835a67114953c2b009f74bd9f1/developer/js/source/lexical-model-compiler/build-trie.ts#L284

File: common/models/templates/src/trie-model.ts [hunk 7]
```diff
@@ -161,6 +161,116 @@
...
       return this.getLastWord(context.left);
     }
+
+    public getRootTraversal(): LexiconTraversal {
+      return new TrieModel.Traversal(this._trie['root'], '');
+    }
+
+    private static Traversal = class implements LexiconTraversal {
+      prefix: String;
+      root: Node;
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:171]: Isn't this always an arbitrary node? It doesn't necessarily have to be the root node of the trie?

```suggestion
      node: Node;
```

File: common/models/templates/src/trie-model.ts [hunk 8]
```diff
@@ -161,6 +161,116 @@
...
+        let root = this.root;
+
+        if(root.type == 'internal') {
+          for(let i = 0; i < root.values.length; i++) {
+            let entry = root.values[i];
+            let entryNode = root.children[entry];
+
+            // SMP check
+            let charCode = entry.charCodeAt(0);
+            if(charCode >= 0xD800 && charCode <= 0xDBFF) {
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:188]: Maybe factor out into a function?

```suggestion
            if(isHighSurrogate(charCode)) {
```

File: common/models/templates/src/trie-model.ts [hunk 9]
```diff
@@ -161,6 +161,116 @@
...
+          let prefix = this.prefix;
+
+          let children = root.entries.filter(function(entry) {
+            return entry.key != prefix && prefix.length < entry.key.length;
+          })
+
+          for(let i = 0; i < children.length; i++) {
+            let entry = children[i];
+            let key = entry.key;
+            let nodeKey = entry.key[prefix.length]
```
Comment (eddieantonio) [inline common/models/templates/src/trie-model.ts:236]:
```suggestion
          for(let {key} of children) {
            let nodeKey = key[prefix.length];
```

File: common/models/types/index.d.ts [hunk 10]
```diff
@@ -17,6 +17,37 @@
...
+
+  /**
+   * Any entries directly keyed by the currently-represented lookup prefix.  Entries and
+   * children may exist simultaneously, but `entries` must always exist when no children are
+   * available in the returned `children()` iterable.
+   * 
+   * For example, with a search prefix of 'the', entries should return ['the'] even if 'then'
+   * and 'there' also exist within the lexicon.
+   */
+  entries: USVString[];
```
Comment (eddieantonio) [inline common/models/types/index.d.ts:48]: This type is incomplete; the implementation above may return `undefined` if there are no matching entries. Honestly, I'd rather this return an empty array rather than `undefined` in the case of no matching entries; it will simplify the implementation of `entries()`, and not drastically affect the check in the caller of this code, e.g., checking for `entries.length > 0` instead of `entriess !== undefined`.

File: common/models/templates/src/trie-model.ts (hunk 1 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:186]: So the check is (mostly) right, but it's the nomenclature that's wrong?

File: common/models/templates/src/trie-model.ts (hunk 7 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:171]: Yes, it is an arbitrary node... the root of its sub-trie.  Parent nodes of the instance stored in `root` can't be revisited from a `LexiconTraversal` instance.

Perhaps documenting that more explicitly is in order.

File: common/models/templates/src/trie-model.ts (hunk 2 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:193]: Is that a TODO for the lm-layer or for the model compiler?  Or both?

File: common/models/types/index.d.ts (hunk 4 above)
Comment (jahorton) [reply common/models/types/index.d.ts:34]: > I think it's saying that, if the traversal is active for th, then char will be "a" or "e". I'm not sure how to succinctly describe this...

Neither do I.  That's why the description is what it is, and why it's hard to follow.

> Another thing: I know USVString implies it

It does?  I've never really been clear on why we made that typealias.

> but could you please note that char can be an astral code points may, with a JavasScript .length of 2?

So, "code point", so may be composed of two "code units".

File: common/models/templates/src/trie-model.ts (hunk 2 above)
Comment (eddieantonio) [reply common/models/templates/src/trie-model.ts:199]: This is a TODO for the model compiler, mostly!

File: common/models/templates/src/trie-model.ts (hunk 1 above)
Comment (eddieantonio) [reply common/models/templates/src/trie-model.ts]: Correct! Just use "non-BMP" or "astral code points" and it's all good :)

File: common/models/templates/src/common.ts [hunk 11]
```diff
@@ -34,4 +36,28 @@ namespace models {
...
+    if(typeof codeUnit == 'string') {
+      codeUnit = codeUnit.charCodeAt(0);
+    }
+
+    return codeUnit >= 0xD800 && codeUnit <= 0xDBFF;
+  }
+
+  export function isSentinel(char: string): boolean {
+    return char == models.SENTINEL_CODE_UNIT;
+  }
```
Comment (eddieantonio) [inline common/models/templates/src/common.ts:62]: 👍  these are very nice!
